File: singletonpattern/singletonpattern.js
class DBConnection1 {
  constructor() {
    if (DBConnection1._conObject) {
      throw new Error("Use DBConnection1.getInstance()");
    }
    // initialization code
  }

  static _instance = new DBConnection1(); // Eagerly initialized

  static getInstance() {
    return DBConnection1._instance;
  }
}

class DBConnection2 {
  constructor() {
    if (DBConnection2._instance) {
      throw new Error("Use DBConnection2.getInstance()");
    }
    // Initialize only once
    DBConnection2._instance = this;
  }

  static getInstance() {
    if (!DBConnection2._instance) {
      DBConnection2._instance = new DBConnection2();
    }
    return DBConnection2._instance;
  }
}

class DBConnection4 {
  constructor() {
    if (DBConnection4._conObject) {
      return DBConnection4._conObject;
    }
    DBConnection4._conObject = this;
    DBConnection4._lock = false
  }

  static getInstance() {
    if (!DBConnection4._conObject) {
      if (!DBConnection4._lock) {
        DBConnection4._lock = true;
        DBConnection4._conObject = new DBConnection4();
      }
    }
    return DBConnection4._conObject;
  }
}

// TEST: DBConnection1 class
const eager1 = DBConnection1.getInstance();
const eager2 = DBConnection1.getInstance();
console.log(`Eager Instance ID: ${eager1 === eager2}`); // true

// TEST: DBConnection2 class
const lazy1 = DBConnection2.getInstance();
const lazy2 = DBConnection2.getInstance();
console.log(`Lazy Instance ID: ${lazy1 === lazy2}`); // true
// new DBConnection(); // ❌ Throws error

// TEST: DBConnection4 class
const threadSafe1 = DBConnection4.getInstance();
const threadSafe2 = DBConnection4.getInstance();
console.log(`Thread-Safe Instance ID: ${threadSafe1 === threadSafe2}`); // true

File: singletonpattern/singletonpattern.py
# 1. EAGER INITIALIZATION METHOD
class DBConnection_1:
    _con_object = object()

    @staticmethod
    def get_instance():
        return DBConnection_1._con_object
    
# 2. LAZY INITIALIZATION METHOD
class DBConnection_2:
    _con_object = None

    def __init__(self):
        pass
    
    @staticmethod
    def get_instance():
        if DBConnection_2._con_object is None:
            DBConnection_2._con_object = DBConnection_2()
        return DBConnection_2._con_object

# 3. LAZY INITIALIZATION WITH THREAD SAFETY (DOUBLE LOCKING)
import threading

# OPTION-I
class DBConnection_3:
    _con_object = None
    _lock = threading.Lock()

    def __new__(cls):
        if cls._con_object is None:
            with cls._lock:
                if cls._con_object is None:
                    cls._con_object = super(DBConnection_3, cls).__new__(cls)
        return cls._con_object

    def __init__(self):
        pass

    @staticmethod
    def get_instance():
        return DBConnection_3._con_object

# OPTION-II
class DBConnection_4:
    _con_object = None
    _lock = threading.Lock()

    def __init__(self):
        pass

    @staticmethod
    def get_instance():
        if DBConnection_4._con_object is None:
            with DBConnection_4._lock:
                if DBConnection_4._con_object is None:
                    DBConnection_4._con_object = DBConnection_4()
        return DBConnection_4._con_object

if __name__ == "__main__":
    con_object_1 = DBConnection_1.get_instance()
    con_object_2 = DBConnection_1.get_instance()
    print(f"Instance ID: {id(con_object_1)}")
    print(f"Instance ID: {id(con_object_2)}")

    con_object_1 = DBConnection_2.get_instance()
    con_object_2 = DBConnection_2.get_instance()
    
    print(f"Instance ID: {id(con_object_1)}")
    print(f"Instance ID: {id(con_object_2)}")

    def test_singleton_3():
        singleton = DBConnection_3.get_instance()
        print(f"Instance ID: {id(singleton)}")
    
    def test_singleton_4():
        singleton = DBConnection_4.get_instance()
        print(f"Instance ID: {id(singleton)}")

     # Create multiple threads to test thread safety
    threads = []
    for i in range(5):
        thread = threading.Thread(target=test_singleton_3)
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()
    
    threads = []
    for i in range(5):
        thread = threading.Thread(target=test_singleton_4)
        threads.append(thread)
        thread.start()

    for thread in threads:
        thread.join()
        